### PR TITLE
OAK-10527 Improve readability of the explain query output

### PR DESCRIPTION
**Pull request by thomasmueller**
I had a to change a lot of tests, because they depended (too much) on implementation details.

Even thought this would be a good idea, I did not switch to Json output so far, because:
* some upstream UI tools depend on the query plans to have a certain form, which would break,
* it would make the patch even larger,
* the Lucene query string might contain double quotes, which would then be escaped, and so the output hard to read

So basically, first the upstream UI tools would 

The Elasticsearch query is Json, which wouldn't need to be escaped if we add it as a child object.

### CODE REVIEW

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        for (String term : new String[] {
+                "union ", "from ", "where ", "and ", "or ", "order by ", "option\\("
+        }) {
+            // SQL is case insensitive, so we use (?i)
+            query = query.replaceAll("(?i) (" + term + ")", "\n  $1");
+        }
+        // remove duplicate newlines
+        query = query.replaceAll("\n+", "\n");
+        return query;
+    }
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: Some of these keywords are common enough that they will likely appear in the text of some queries, which will then result in line breaks being added in the wrong places. Would it be very complicated to change the logic to detect if they words are used as sql keywords or not? Maybe something with checking if the word appears inside quotes or not? Or maybe it's ok for the output to sometimes be wrongly formatted, as this is mostly a visual aid.

**Review by nfsantos — COMMENTED** [review]
I did not look at the tests, just have a minor comment.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 1 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: Yes this is a concern, if the output is then run as a query... (By the human or automatically).
I was thinking on how to solve it, but I don't think there is a simple solution that would cover all cases. (Basically, don't use regular expressions.) Maybe we could add a comment in the output, but that's ugly as well...

The output is only used in one place: LOG.debug, so I think the risk is not all that great now.
My recommendation is to just accept that risk.
But not use the code anywhere else, that's why I added the warning in the Javadoc.

Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: Well, actually, I implemented the slightly more complex solution now (including randomized test, which is anyway good to have).

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    private static String formatSQL(String query) {
+        StringBuilder buff = new StringBuilder(query);
+        for (int i = 0; i < buff.length(); i++) {
+            char c = buff.charAt(i);
+            if (c == '\'' || c == '"') {
+                while (++i < buff.length() && buff.charAt(i) != c) {
+                    // skip
+                }
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:126]: Won't this fail if there is some string in the query that contains " inside it? For instance, a string like `"he said \"Hello\""`.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/nodetype/NodeTypeIndex.java [hunk 3]
```diff
@@ -102,7 +102,10 @@ public Cursor query(Filter filter, NodeState root) {
     
     @Override
     public String getPlan(Filter filter, NodeState root) {
-        return "nodeType " + filter.toString();
+        return "nodeType\n" +
+                "    path: " + filter.getPath() + "\n" +
+                "    primaryTypes: " + filter.getPrimaryTypes().toString() + "\n" +
+                "    mixinTypes: " + filter.getMixinTypes().toString() + "\n";
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/nodetype/NodeTypeIndex.java]: `toString` calls are redundant:

```suggestion
                "    primaryTypes: " + filter.getPrimaryTypes() + "\n" +
                "    mixinTypes: " + filter.getMixinTypes() + "\n";
```

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    public static boolean isXPath(String query, String language) {
+        if (language != null) {
+            return "xpath".equals(language);
+        }
+        query = query.trim().toLowerCase(Locale.ENGLISH);
+        // explain queries
+        if (query.startsWith("explain")) {
+            query = query.substring("explain".length()).trim();
+            if (query.startsWith("measure")) {
+                query = query.substring("explain".length()).trim();
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: should it be like this? can we add a test for this case?
```suggestion
                query = query.substring("measure".length()).trim();
```

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            }
+        }
+        // union queries
+        while (query.startsWith("(")) {
+            query = query.substring("(".length()).trim();
+        }
+        if (query.startsWith("select")) {
+            return false;
+        }
+        return true;
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:64]: this could be simplified 

```suggestion
        return !query.startsWith("select");
```

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    /**
+     * Format the query into a more human-readable way, by adding newlines.
+     * Warning: newlines are also added inside e.g. string literals.
+     *
+     * @param query the query (may not be null)
+     * @param language the query language, or null if unknown
+     * @return the formatted query
+     */
+    public static String format(String query, String language) {
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: since the `query` param cannot be `null`, what about annotating it?
```suggestion
    public static String format(@NotNull String query, String language) {
```

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/SelectorImpl.java [hunk 7]
```diff
@@ -427,10 +426,14 @@ public String getPlan(NodeState rootState) {
                 buff.append(index.getPlan(createFilter(true), rootState));
             }
         } else {
-            buff.append("no-index");
+            buff.append("no-index\n");
         }
         if (!selectorConstraints.isEmpty()) {
```
Comment (fabriziofortino) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/SelectorImpl.java:431]: this `if` block is empty. Should we remove it?

File: oak-core/src/test/java/org/apache/jackrabbit/oak/query/QueryFormatterTest.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+ */
+package org.apache.jackrabbit.oak.query;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.Random;
+
+import org.apache.jackrabbit.oak.query.QueryFormatter;
```
Comment (fabriziofortino) [inline oak-core/src/test/java/org/apache/jackrabbit/oak/query/QueryFormatterTest.java]: not needed
```suggestion
```

**Review by fabriziofortino — CHANGES_REQUESTED** [review]

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 2 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:126]: Backslash is not an escape character. The escape character for ' is ', so that it would be:

    "he said ""Hello"""

and

    'he said ''Hello'''

Because of this, we just need to search for the next ' / ". The "" and '' are correctly recognised. There's also a test for that.

Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:126]: See the docs at https://jackrabbit.apache.org/oak/docs/query/grammar-sql2.html#static_operand and https://jackrabbit.apache.org/oak/docs/query/grammar-xpath.html#static_operand

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 5 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:64]: Actually... I think the current code is clearer in this case: The "not" can be confusing.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/SelectorImpl.java (hunk 7 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/SelectorImpl.java:431]: Well, it is not empty: I would like to keep the comment and the commented code...

**Review by fabriziofortino — APPROVED** [review]

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Formatter for JCR queries in order to make them easier to read. Formatting is
+ * done on a best-effort basis.
+ *
+ * Warning: Care was taken to not add newlines inside string literals and so on,
+ * but there is still no guarantee that the formatted query is semantically
+ * equal to the original one. It is recommended to run queries that are returned
+ * by these methods.
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: Don't you mean that it is recommended **not** to run queries...?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * Detect whether the query is an XPath query.
+     *
+     * @param query the query
+     * @param language the language, if known, or null
+     * @return true if xpath
+     */
+    public static boolean isXPath(String query, String language) {
+        if (language != null) {
+            return "xpath".equals(language);
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:46]: Being case sensitive is intentional? The query string is treated as case insensitive. If the language is supposed to be case sensitive, maybe add a note about it in the comments.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        query = query.trim().toLowerCase(Locale.ENGLISH);
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:49]: If we do not want any locale related interpretation, it's better to use [Locale.ROOT](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#ROOT).

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        // explain queries
+        if (query.startsWith("explain")) {
+            query = query.substring("explain".length()).trim();
+            if (query.startsWith("measure")) {
+                query = query.substring("measure".length()).trim();
+            }
+        }
+        // union queries
+        while (query.startsWith("(")) {
+            query = query.substring("(".length()).trim();
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:59]: Maybe replace with `substring(1)` since we are just removing the first character? I understand the reason for this convention when it is a word, prevents mistakes, but when it's only a single character, maybe it is simpler to use 1.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        }
+        if (query.startsWith("select")) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Format the query into a more human-readable way, by adding newlines.
+     * Warning: newlines are also added inside e.g. string literals.
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]:
```suggestion
     * Warning: newlines are also added inside string literals.
```
In what conditions? Do you mean that they may also be added in some rare cases?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+            query = query.substring("(".length()).trim();
+        }
+        if (query.startsWith("select")) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Format the query into a more human-readable way, by adding newlines.
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]:
```suggestion
     * Format the query into a more human-readable way by adding newlines.
```
Adding newlines where? This sentence could be a bit more clear.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+     * @param query the query (may not be null)
+     * @param language the query language, or null if unknown
+     * @return the formatted query
+     */
+    public static String format(@NotNull String query, String language) {
+        boolean xpath = isXPath(query, language);
+        if (xpath) {
+            return formatXPath(query);
+        } else {
+            return formatSQL(query);
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:80]: We are not sure here if the language is sql. If language is specified, add a check that it is either sql or xpath. If we autodetect, then can we check if it is sql by looking at if it starts with `select`? That way, we could have a nice error message saying that we do not recognize the language instead of failing with a syntax error or returning a string with newlines added in random places.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        }
+    }
+
+    private static String formatXPath(String query) {
+        StringBuilder buff = new StringBuilder(query);
+        for (int i = 0; i < buff.length(); i++) {
+            char c = buff.charAt(i);
+            if (c == '\'' || c == '"') {
```
Comment (nfsantos) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:88]: This could be transformed into a switch.

File: oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/property/MultiPropertyOrTest.java [hunk 17]
```diff
@@ -121,22 +121,17 @@ public void unionSortResultCount() throws Exception {
         root.commit();
 
         List<Integer> nodes = Lists.newArrayList();
-        Random r = new Random();
-        int seed = -2;
+        Random r = new Random(1);
```
Comment (nfsantos) [inline oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/property/MultiPropertyOrTest.java:124]: Why? You want the test to be deterministic?

**Review by nfsantos — COMMENTED** [review]
Just a few more comments.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 9 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java]: Of course 😵‍💫

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 10 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:46]: The _language_ is case sensitive. The query, for SQL, is not (for XPath it is).

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 11 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:49]: Locale.ENGLISH is safe according to https://stackoverflow.com/questions/10336730/which-locale-should-i-specify-when-i-call-stringtolowercase . We use that in other parts too, so I would keep this.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 12 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:59]: I used this for consistency, and to prevent errors in maintenance.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 15 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:80]: > we could have a nice error message saying that we do not recognize the language

I don't see a need for this currently. The current use case is debug log, where we actually always know which language it is.

I do want to re-use this in the near future for http://oakutils.appspot.com/generate/index where the language detection is (unfortunately) incorrect.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java (hunk 16 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java:88]: Yes. But that wouldn't necessarily be better. I prefer the "if else" in this case.

File: oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/property/MultiPropertyOrTest.java (hunk 17 above)
Comment (thomasmueller) [reply oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/property/MultiPropertyOrTest.java:124]: Yes.

**Review by nfsantos — APPROVED** [review]

**Review by amit-jain — APPROVED** [review]
LGTM. Haven't looked at the tests carefully though.
